cranelift-debug/src/write_debuginfo.rs

```diff
@@ -0,0 +1,209 @@
+//! Writing native DWARF sections.
+
+use crate::transform::TransformedDwarf;
+
+use gimli::write::{
+    Address, DebugAbbrev, DebugInfo, DebugLine, DebugLineStr, DebugRanges, DebugRngLists, DebugStr,
+    EndianVec, Result, SectionId, Sections, Writer,
+};
+use gimli::RunTimeEndian;
+
+use faerie::artifact::Decl;
+use faerie::*;
+
+struct DebugReloc {
+    offset: u32,
+    size: u8,
+    name: String,
+    addend: i64,
+}
+
+macro_rules! decl_section {
+    ($artifact:ident . $section:ident = $name:expr) => {
+        $artifact
+            .declare_with(
+                SectionId::$section.name(),
+                Decl::debug_section(),
+                $name.0.writer.into_vec(),
+            )
+            .unwrap();
+    };
+}
+
+macro_rules! sect_relocs {
+    ($artifact:ident . $section:ident = $name:expr) => {
+        for reloc in $name.0.relocs {
+            $artifact
+                .link_with(
+                    faerie::Link {
+                        from: SectionId::$section.name(),
+                        to: &reloc.name,
+                        at: u64::from(reloc.offset),
+                    },
+                    faerie::Reloc::Debug {
+                        size: reloc.size,
+                        addend: reloc.addend as i32,
+                    },
+                )
+                .expect("faerie relocation error");
+        }
+    };
+}
+
+pub enum ResolvedSymbol {
+    PhysicalAddress(u64),
+    Reloc { name: String, addend: i64 },
+}
+
+pub trait SymbolResolver {
+    fn resolve_symbol(&self, symbol: usize, addend: i64) -> ResolvedSymbol;
+}
+
+pub fn emit_dwarf(
+    artifact: &mut Artifact,
+    mut dwarf: TransformedDwarf,
+    symbol_resolver: &SymbolResolver,
+) {
```

Comment:
With latest gimli this can be simplified to:

```rust
let mut sections = Sections::new(WriterRelocate::new(endian, symbol_resolver));
dwarf.write(&mut sections).unwrap();

let _: Result<()> = sections.for_each_mut(|id, section| {
    if !section.writer.slice().is_empty() {
        artifact
   .declare_with(id.name(), Decl::debug_section(), section.writer.take())
   .unwrap();
    }
    Ok(())
});

let _: Result<()> = sections.for_each(|id, section| {
    for reloc in &section.relocs {
        artifact
            .link_with(
                faerie::Link {
                    from: id.name(),
                    to: reloc.name(self),
                    at: u64::from(reloc.offset),
                },
                faerie::Reloc::Debug {
                    size: reloc.size,
                    addend: reloc.addend as i32,
                },
            )
            .expect("faerie relocation error");
    }
    Ok(())
});
```

(adapted from https://github.com/bjorn3/rustc_codegen_cranelift/blob/47b1ef24e48bd3c250d14c756eb5e9de6c1ece20/src/debuginfo.rs#L188-L217)